executor/mem_reader.go

```diff
@@ -305,3 +298,100 @@ func reverseDatumSlice(rows [][]types.Datum) {
 		rows[i], rows[j] = rows[j], rows[i]
 	}
 }
+
+func (m *memIndexReader) getMemRowsHandle() ([]int64, error) {
+	pkTp := types.NewFieldType(mysql.TypeLonglong)
+	if m.table.PKIsHandle {
+		for _, col := range m.table.Columns {
+			if mysql.HasPriKeyFlag(col.Flag) {
+				pkTp = &col.FieldType
+				break
+			}
+		}
+	}
+	handles := make([]int64, 0, m.addedRowsLen)
+	err := iterTxnMemBuffer(m.ctx, m.kvRanges, func(key, value []byte) error {
+		handle, err := tablecodec.DecodeIndexHandle(key, value, len(m.index.Columns), pkTp)
+		if err != nil {
+			return err
+		}
+		handles = append(handles, handle)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if m.desc {
+		for i, j := 0, len(handles)-1; i < j; i, j = i+1, j-1 {
+			handles[i], handles[j] = handles[j], handles[i]
+		}
+	}
+	return handles, nil
+}
+
+type memIndexLookUpReader struct {
+	ctx           sessionctx.Context
+	index         *model.IndexInfo
+	columns       []*model.ColumnInfo
+	table         table.Table
+	desc          bool
+	conditions    []expression.Expression
+	retFieldTypes []*types.FieldType
+
+	idxReader *memIndexReader
+}
+
+func buildMemIndexLookUpReader(us *UnionScanExec, idxLookUpReader *IndexLookUpExecutor) *memIndexLookUpReader {
+	kvRanges := idxLookUpReader.kvRanges
+	outputOffset := []int{len(idxLookUpReader.index.Columns)}
+	memIdxReader := &memIndexReader{
+		ctx:              us.ctx,
+		index:            idxLookUpReader.index,
+		table:            idxLookUpReader.table.Meta(),
+		kvRanges:         kvRanges,
+		desc:             idxLookUpReader.desc,
+		addedRowsLen:     len(us.dirty.addedRows),
+		retFieldTypes:    retTypes(us),
+		outputOffset:     outputOffset,
+		belowHandleIndex: us.belowHandleIndex,
+	}
+
+	return &memIndexLookUpReader{
```

Comment:
Why not embed `memIndexReader` to `memIndexLookUpReader`?

Reply:
`PhysicalIndexLookUpReader` also use `indexPlan` and `tablePlan` separately. 
And I think it is more clear.